spec/lang/intrinsic.md

```diff
@@ -194,21 +194,33 @@ impl<M: Memory> Machine<M> {
         };
 
         let func = self.fn_from_addr(ptr.addr)?;
+        if func.args.len() != 1 {
+            throw_ub!("invalid first argument to `Intrinsic::Spawn`, function should take one argument.");
+        }
 
-        if func.args.len() != 0 {
-            throw_ub!("invalid first argument to `Intrinsic::Spawn`, function takes arguments");
+        let (data_ptr, data_ptr_ty) = arguments[1];
+        if !matches!(data_ptr_ty, Type::Ptr(PtrType::Raw)) {
+            throw_ub!("invalid second argument to `Intrinsic::Spawn`, data pointer should be a raw pointer.");
+        }
+        // This logic is taken from `eval(Terminator::Call)` and `eval_argument`.
+        let data_ptr_pty = PlaceType::new(data_ptr_ty, M::T::PTR_ALIGN);
+        if !Self::check_abi_compatibility(data_ptr_pty, func.locals[func.args[0]]) {
+            throw_ub!("invalid first argument to `Intrinsic::Spawn`, function should take a pointer as an argument.");
         }
 
-        // This is taken from Miri. It also does not allow for a return value in the root function of a thread.
-        if func.ret.is_some() {
-            throw_ub!("invalid first argument to `Intrinsic::Spawn`, function returns something");
+        // This is taken from Miri. It discards the return value of the function.
+        // We enforce this by only allowing functions that return () and !.
+        if let Some(name) = func.ret {
+            if func.locals[name].ty.size::<M::T>() != Size::ZERO {
+                throw_ub!("invalid first argument to `Intrinsic::Spawn`, function returns something non zero sized");
+            }
         }
```

Comment:
This check isn't actually needed any more I think. Now that we allocate a proper return place in `spawn`, we can accept arbitrary return types.

Reply:
Doesn't this suggest that we do something with the return value?

Reply:
Why would it suggest that? We just discard the return value. I don't think we have to reject that.